File: 2021/25.py
m = [line[:-1] for line in open("25.txt").readlines()]
moving = True
r = 0
while moving:
    #print(f"round {r}")
    #print('\n'.join(m))
    r += 1
    n = ['.' * len(m[i]) for i in range(len(m))]
    moving = False
    for i in range(len(m)):
        for j in range(len(m[i])):
            if m[i][j] == '>' and j < len(m[i])-1 and m[i][j+1] == '.':
                n[i] = n[i][:j] + '.>' + n[i][j+2:]
                moving = True
            elif m[i][j] == '>' and j == len(m[i])-1 and m[i][0] == '.':
                n[i] = '>' + n[i][1:j] + '.'
                moving = True
            
            elif m[i][j] == '>':
                n[i] = n[i][:j] + '>' + n[i][j+1:]
            elif m[i][j] == 'v':
                n[i] = n[i][:j] + 'v' + n[i][j+1:]

    m = n.copy()
    for i in range(len(m)):
        for j in range(len(m[i])):
            if m[i][j] == 'v' and i < len(m)-1 and m[i+1][j] == '.':
                n[i] = n[i][:j] + '.' + n[i][j+1:]
                n[i+1] = n[i+1][:j] + 'v' + n[i+1][j+1:]
                moving = True
            elif m[i][j] == 'v' and i == len(m)-1 and m[0][j] == '.':
                n[i] = n[i][:j] + '.' + n[i][j+1:]
                n[0] = n[0][:j] + 'v' + n[0][j+1:]
                moving = True
            elif m[i][j] == 'v':
                n[i] = n[i][:j] + 'v' + n[i][j+1:]
    
    m = n

print(r)

File: 2022/25.jl
function decode(snafu)

    total = 0
    mult = 1    
    for i in range(length(snafu), 1, step=-1)
        dig = undef
        if isdigit(snafu[i])
            dig = parse(Int, snafu[i])
        elseif snafu[i] == '-'
            dig = -1
        else
            dig = -2
        end
        total += mult * dig
        mult *= 5
    end
    return total
end

function encode(decimal)

    snafu = ""
    mult = 1
    while decimal > 0
        rest = decimal % (mult * 5)
        dig = rest ÷ mult
        if dig < 3
            snafu = string(dig) * snafu
            decimal -= dig * mult
        elseif dig == 3
            snafu = "=" * snafu
            decimal += 2 * mult
        elseif dig == 4
            snafu = "-" * snafu
            decimal += 1 * mult
        end
        mult *= 5
    end
    return snafu
end

function fully_of_hot_air()

    total = 0
    for line in readlines("25.txt")
        decimal = decode(line)
        total += decimal
    end
    println(total)
    println(encode(total))
end

@time fully_of_hot_air()